File: node/pkg/fetcher/reducer.go
package fetcher

import (
	"fmt"
	"math"
)

type Definition struct {
	Url      string            `json:"url"`
	Headers  map[string]string `json:"headers"`
	Method   string            `json:"method"`
	Reducers []Reducer         `json:"reducers"`
}

type Reducer struct {
	Function string      `json:"function"`
	Args     interface{} `json:"args"`
}

func ReduceAll(raw interface{}, reducers []Reducer) (float64, error) {
	var result float64
	for _, reducer := range reducers {
		var err error
		raw, err = reduce(raw, reducer)
		if err != nil {
			return 0, err
		}
	}
	result, ok := raw.(float64)
	if !ok {
		return 0, fmt.Errorf("cannot cast raw data to float")
	}
	return result, nil
}

func reduce(raw interface{}, reducer Reducer) (interface{}, error) {
	switch reducer.Function {
	case "INDEX":
		castedRaw, ok := raw.([]interface{})
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to []interface{}")
		}
		return castedRaw[reducer.Args.(int)], nil
	case "PARSE", "PATH":
		castedRaw, ok := raw.(map[string]interface{})
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to map[string]interface{}")
		}
		args := reducer.Args.([]string)
		for _, arg := range args {
			castedRaw = castedRaw[arg].(map[string]interface{})
		}
		return castedRaw, nil
	case "MUL":
		castedRaw, ok := raw.(float64)
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to float")
		}
		return castedRaw * reducer.Args.(float64), nil
	case "POW10":
		castedRaw, ok := raw.(float64)
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to float")
		}
		return float64(math.Pow10(reducer.Args.(int))) * castedRaw, nil
	case "ROUND":
		castedRaw, ok := raw.(float64)
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to float")
		}
		return math.Round(castedRaw), nil
	case "DIV":
		castedRaw, ok := raw.(float64)
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to float")
		}
		return castedRaw / reducer.Args.(float64), nil
	case "DIVFROM":
		castedRaw, ok := raw.(float64)
		if !ok {
			return nil, fmt.Errorf("cannot cast raw data to float")
		}
		return reducer.Args.(float64) / castedRaw, nil
	default:
		return nil, fmt.Errorf("unknown reducer function: %s", reducer.Function)
	}

}

File: util/src/reducer.ts
import { ReducerError, ReducerErrorCode } from './errors'

export const REDUCER_MAPPING = {
  PATH: parseFn,
  PARSE: parseFn,
  MUL: mulFn,
  POW10: pow10Fn,
  ROUND: roundFn,
  INDEX: indexFn,
  DIV: divFn,
  DIVFROM: divFromFn,
}

/**
 * Access data in JSON based on given path.
 *
 * Example
 * let obj = {
 *     RAW: { ETH: { USD: { PRICE: 123 } } },
 *     DISPLAY: { ETH: { USD: [Object] } }
 * }
 * const fn = parseFn(['RAW', 'ETH', 'USD', 'PRICE'])
 * fn(obj) // return 123
 */

export function parseFn(args: string | string[]) {
  if (typeof args == 'string') {
    args = args.split(',')
  }

  function wrapper(obj) {
    for (const a of args) {
      if (a in obj) obj = obj[a]
      else throw new ReducerError(ReducerErrorCode.MissingKeyInJson)
    }
    return obj
  }
  return wrapper
}

export function mulFn(args: number) {
  function wrapper(value: number) {
    return value * args
  }
  return wrapper
}

export function divFn(args: number) {
  function wrapper(value: number) {
    return value / args
  }
  return wrapper
}

export function divFromFn(args: number) {
  function wrapper(value: number) {
    if (value == 0) {
      throw new ReducerError(ReducerErrorCode.DivisionByZero)
    }
    return args / value
  }
  return wrapper
}

export function pow10Fn(args: number) {
  function wrapper(value: number) {
    return Number(Math.pow(10, args)) * value
  }
  return wrapper
}

export function roundFn() {
  function wrapper(value: number) {
    return Math.round(value)
  }
  return wrapper
}

export function indexFn(args: number) {
  if (args < 0) {
    throw new ReducerError(ReducerErrorCode.IndexOutOfBoundaries)
  }

  function wrapper(obj) {
    if (args >= obj.length) {
      throw new ReducerError(ReducerErrorCode.IndexOutOfBoundaries)
    } else {
      return obj[args]
    }
  }
  return wrapper
}

export function buildReducer(reducerMapping, reducers) {
  return reducers.map((r) => {
    const reducer = reducerMapping[r.function.toUpperCase()]
    if (!reducer) {
      throw new ReducerError(ReducerErrorCode.InvalidReducer)
    }
    return reducer(r?.args)
  })
}

// https://medium.com/javascript-scene/reduce-composing-software-fe22f0c39a1d
export const pipe =
  (...fns) =>
  (x) =>
    fns.reduce((v, f) => f(v), x)

export function checkDataFormat(data) {
  if (!data) {
    throw new ReducerError(ReducerErrorCode.InvalidData)
  } else if (!Number.isInteger(data)) {
    throw new ReducerError(ReducerErrorCode.InvalidDataFormat)
  }
}
